File: examples/small_examples.cpp
#ifdef _WIN32
// enable exception handling for windows
// this requires 'int main(int, char**)' function definition
// therefore 'int main()' is dissabled
#define CFCG_EXCEPTION_HANDLING
#endif

#include "BigIntegerLibrary.hh"

#include "utils.h"
#include <iomanip> // set precision

int main() {
    // read ant string and print it
    const std::string antStr = cf::readAntString(CHAOS_FILE_PATH "Ant_10.ant");
    std::cout << "Ant string:\t" << antStr << '\n' << std::endl;

    // iterate through some colors
    // option 1:
    std::cout << "Iterate throgh colors, option 1:\n";
    const auto colors = {cf::Color::GREEN, cf::Color::RandomColor()};
    for (const auto& e : colors)
        std::cout << e << "\t\t";
    std::cout << "\n\n";

    // option 2:
    std::cout << "Iterate throgh colors, option 2:\n";

    // use as many colors as you like ;)
    auto iter = cf::Color::CreateEndlessColorIterator(cf::Color::BLACK, cf::Color::RandomColor());
    for (int i = 0; i < 3; ++i) {
        std::cout << *iter << '\n';
        ++iter;
    }
    std::cout << std::endl;

    // read pal file and print a color
    std::vector<cf::Color> pal = cf::readPaletteFromFile(CHAOS_FILE_PATH "Chaos_ant.pal");
    std::cout << "Palette at index 4:\t\t" << pal[4] << std::endl;
    std::cout << "Palette at index 4 inverted:\t" << pal[4].invert() << std::endl << std::endl;

    // read dat file and print a vector
    std::vector<glm::vec3> dat = cf::readDATFile(CHAOS_FILE_PATH "Life1.dat");
    std::cout << "Dat file at index 0:\t" << dat[0] << std::endl << std::endl;

    // calculation with very very large ints :)
    BigInteger veryBigNumber = stringToBigInteger("999999999999999999999999999999999");
    veryBigNumber *= veryBigNumber;
    veryBigNumber /= 10;
    veryBigNumber += BigInteger(111111111111111111) + 1;
    std::cout << "Very big number calculation result:\n" << bigIntegerToString(veryBigNumber) << std::endl << std::endl;

    // convert between degrees/radians
    // Note: OpenGL uses degree, glm radians (see 'window_3D_example.cpp')
    const float deg = cf::degree2radian(180.f);
    const float rad = cf::radian2degree(glm::pi<float>() / 2.f);
    std::cout << "180 degree in radian:  " << deg << std::endl << " Pi half   in degree:  " << rad << std::endl << std::endl;

    // print floating point values with different precisions
    // based on: http://www.cplusplus.com/reference/iomanip/setprecision/
    const double pi_approx = 3.14159;
    std::cout << "Different pi approximation outputs:\n"
              << std::setprecision(5) << pi_approx << "\t\t" << std::setprecision(9) << pi_approx << "\t\t" << std::fixed
              << std::setprecision(5) << pi_approx << "\t\t" << std::setprecision(9) << pi_approx << std::endl
              << std::endl;

    // Direction example
    cf::Direction::AbsoluteDirection west = cf::Direction::AbsoluteDirection::WEST;
    std::cout << "If your current direction is 'WEST'\nand you take a right turn your new direction will be:\t'"
              << cf::Direction::toString(cf::Direction::getNextiDirection(west, cf::Direction::RelativeDirection::RIGHT))
              << "'\n"
              << std::endl;

    // change console color
    std::cout << termcolor::blue << "blue " << termcolor::bold << "bold " << termcolor::on_cyan << "onGrey "
              << termcolor::reverse << "reverse " << termcolor::reset << std::endl;

    // program end
    std::cout << "Press enter to end the process" << std::flush;
    cf::Console::waitKey();
    return 0;
}

File: python/small_examples.py
# small_examples.py

import math
from libcfcg import cf

# read 'ant' files
value = cf.readAntString('chaos_files/Ant_10.ant')
print("String in ant file: ", value)

# iterate over multiple colors
colors = [ cf.Color.RED, cf.Color.GREEN, cf.Color.BLUE, cf.Color.ORANGE ]
for color in colors:
    print("Current color: ", color.r, color.g, color.b)

# calculation with big numbers is built into python
value = 12345678901234567890
print(value, " modulo 13 is: ", value % 13)
value = 2**1024 # 2 to the power of 1024
print(value, " modulo 13 is: ", value % 13)

# convert between degree and radians
value = 180.0
print(value, " in radiants is: ", cf.degree2radian(value))
value = math.pi * 1.5
print(value, " in degree is: ", cf.radian2degree(value))

# direction example
west = cf.Direction.AbsoluteDirection_WEST
right = cf.Direction.RelativeDirection_RIGHT
westTurnRight = cf.Direction.getNextiDirection(west, right)
print("If your current direction is ",
        cf.Direction_toString(west), "\n",
        "and you take a right turn you are looking: ",
        cf.Direction_toString(westTurnRight))

# read integer/float values
value1 = float(input('Enter a float value: '))
value2 = int(input('Enter an int value: '))
print("float value:", value1, "   int value:", value2)

window = None # destroy window
